Marlin/src/HAL/HAL_DUE/usb/usb_task.c

```diff
@@ -178,9 +124,183 @@ void usb_task_cdc_config(uint8_t port, usb_cdc_line_coding_t * cfg)
 
 void usb_task_cdc_set_dtr(uint8_t port, bool b_enable)
 {
-	if (b_enable) {
-	} else {
-	}
+  if (b_enable) {
+  } else {
+  }
+}
+
+/// Microsoft WCID descriptor
+typedef struct USB_MicrosoftCompatibleDescriptor_Interface {
+  uint8_t bFirstInterfaceNumber;
+  uint8_t reserved1;
+  uint8_t compatibleID[8];
+  uint8_t subCompatibleID[8];
+  uint8_t reserved2[6];
+} __attribute__((packed)) USB_MicrosoftCompatibleDescriptor_Interface;
+
+typedef struct USB_MicrosoftCompatibleDescriptor {
+  uint32_t dwLength;
+  uint16_t bcdVersion;
+  uint16_t wIndex;
+  uint8_t bCount;
+  uint8_t reserved[7];
+  USB_MicrosoftCompatibleDescriptor_Interface interfaces[];
+} __attribute__((packed)) USB_MicrosoftCompatibleDescriptor;
+
+// 3D Printer compatible descriptor
+static USB_MicrosoftCompatibleDescriptor microsoft_compatible_id_descriptor = {
+  .dwLength = sizeof(USB_MicrosoftCompatibleDescriptor) +
+              1*sizeof(USB_MicrosoftCompatibleDescriptor_Interface),
+  .bcdVersion = 0x0100,
+  .wIndex = 0x0004,
+  .bCount = 1,
+  .reserved = {0, 0, 0, 0, 0, 0, 0},
+  .interfaces = {
+    {
+      .bFirstInterfaceNumber = 0,
+      .reserved1 = 1,
+      .compatibleID = "3DPRINT",
+      .subCompatibleID = {0, 0, 0, 0, 0, 0, 0, 0},
+      .reserved2 = {0, 0, 0, 0, 0, 0},
+    }
+  }
+};
+
+#define xstr(s) str(s)
+#define str(s) #s
+
+#define MS3DPRINT_CONFIG      u"MS3DPrintConfig"
+#define MS3DPRINT_CONFIG_DATA \
+  u"Base=SD\0"\
+  u"Job3DOutputAreaWidth=" xstr(X_BED_SIZE) "000\0"\
+  u"Job3DOutputAreaDepth=" xstr(Y_BED_SIZE) "000\0"\
+  u"Job3DOutputAreaHeight=" xstr(Z_MAX_POS) "000\0"\
+  u"filamentdiameter=1750\0"
```

Comment:
> should we be using the same manufacturer in usb descriptors? I used 'marlinfw.org' in the LPC1768.

I'm not sure. I would say go with marlinfw.org or org.marlinfw (depending on which is more appropriate for this type of use) and if needed it can be changed later.

Reply:
> probably shouldn't be hard coded but DEFAULT_NOMINAL_FILAMENT_DIA is not the correct format

Is this string used very much? If not, the string can be copied to SRAM and the value portion replaced when the string is employed.

Reply:
In examining the code, I see that `microsoft_extended_properties_descriptor` is a `static` global, so it can be initialized by code — either an initializer or in `usb_task_other_requests` itself, based on the current filament size setting in the firmware, if so desired.

Reply:
@thinkyhead :You are right. Those descriptors could also be in ROM, but i forgot.. You are right.We can patch in the filament size...